jax/config.py

```diff
@@ -143,14 +149,85 @@ def disable_omnistaging(self):
-    lib.jax_jit.thread_local_state().enable_x64 = bool(state)
-
+  def define_bool_state(self, name: str, default: bool, help: str):
+    """Set up thread-local state and return a contextmanager for managing it.
+
+    This function is a convenience wrapper. It defines a flag and corresponding
+    thread-local state, which can be managed via the contextmanager it returns.
+
+    The thread-local state value can be read via the ``config.<option_name>``
+    attribute, where ``config`` is the singleton ``Config`` instance.
+
+    Args:
+      name: string, converted to lowercase to define the name of the config
+        option (and absl flag). It is converted to uppercase to define the
+        corresponding shell environment variable.
+      default: boolean, a default value for the option.
+      help: string, used to populate the flag help information as well as the
+        docstring of the returned context manager.
+
+    Returns:
+      A contextmanager to control the thread-local state value.
+
+    Example:
+
+      enable_foo = config.define_bool_state(
+          name='jax_enable_foo',
+          default=False,
+          help='Enable foo.')
+
+      # Now the JAX_ENABLE_FOO shell environment variable and --jax_enable_foo
+      # command-line flag can be used to control the process-level value of
+      # the configuration option, in addition to using e.g.
+      # ``config.update("jax_enable_foo", True)`` directly. We can also use a
+      # context manager:
+
+      with enable_foo(True):
+        ...
+
+    Accessing ``config.FLAGS.jax_enable_foo`` is different from accessing the
+    thread-local state value via ``config.jax_enable_foo``: the former reads the
+    flag value determined set by the environment variable or command-line flag
+    and does not read the thread-local state, whereas the latter reads the
+    thread-local state value managed by the contextmanager. Think of the
+    contextmanager state as a layer on top of the flag value: if no
+    contextmanager is in use then ``config.jax_enable_foo`` reflects the flag
+    value ``config.FLAGS.jax_enable_foo``, whereas if a contextmanager is in use
+    then only ``config.jax_enable_foo`` is updated. So in general using
+    ``config.jax_enable_foo`` is best.
+    """
+    name = name.lower()
+    self.DEFINE_bool(name, bool_env(name.upper(), default), help)
+
+    def get_state(self):
+      val = getattr(_thread_local_state, name, unset)
+      return val if val is not unset else self.read(name)
+    setattr(Config, name, property(get_state))
+
+    @contextlib.contextmanager
+    def set_state(new_val: bool):
+      prev_val = getattr(_thread_local_state, name, unset)
+      setattr(_thread_local_state, name, new_val)
+      try:
+        yield
+      finally:
+        if prev_val is unset:
+          delattr(_thread_local_state, name)
+        else:
+          setattr(_thread_local_state, name, prev_val)
+    set_state.__name__ = name.lstrip('jax_')
```

Comment:
Just because I've been bitten by this before, lstrip strips characters, not substrings:
```python
>>> 'jax_array'.lstrip('jax_')
'rray'
```
we probably want something like
```python
set_state.__name__ = name[4:] if name.startswith('jax_') else name
```

Reply:
Aha! You got me. Thanks. (Looks like Python 3.9 has `str.removeprefix`...)